neuralmonkey/decoders/beam_search_decoder.py

```diff
@@ -248,65 +274,119 @@ def body(*args) -> BeamSearchLoopState:
             logprobs = unfinished_logprobs + finished_logprobs
 
             # update hypothesis scores
-            # shape(hyp_probs) = beam x vocabulary
+            # shape(hyp_probs) = (batch*beam) x vocabulary
             hyp_probs = tf.expand_dims(bs_state.logprob_sum, 1) + logprobs
 
             # update hypothesis lengths
             hyp_lengths = bs_state.lengths + 1 - tf.to_int32(bs_state.finished)
 
-            # shape(scores) = beam x vocabulary
+            # shape(scores) = (batch*beam) x vocabulary
             scores = hyp_probs / tf.expand_dims(
                 self._length_penalty(hyp_lengths), 1)
 
-            # flatten so we can use top_k
-            scores_flat = tf.reshape(scores, [-1])
+            # reshape to batch_size x (beam_size*vocabulary) for topk
+            scores_flat = tf.reshape(
+                scores, [-1, input_beam_size * len(self.vocabulary)])
 
-            # shape(both) = beam
+            # shape(both) = batch x beam
             topk_scores, topk_indices = tf.nn.top_k(
                 scores_flat, self._beam_size)
+            topk_indices.set_shape([None, self._beam_size])
+            topk_scores.set_shape([None, self._beam_size])
 
-            topk_scores.set_shape([self._beam_size])
-            topk_indices.set_shape([self._beam_size])
+            # First, we need the offset to flatten topk indices so we can
+            # gather correct hyp_probs values
+            #
+            # ("|" marks beam border, "||" marks batch border)
+            # e.g. scores (vocabulary=2, beam_size=3, batch_size=2):
+            #
+            # [ 2.0, 3.0 | 1.0, 4.0 | 2.0, 5.0 ]
+            # [ 1.0, 2.0 | 3.0, 4.0 | 5.0, 1.0 ]
+            #
+            # topk(k=3) produces following indices:
+            # [ 5, 3, 1 ]
+            # [ 4, 3, 2 ]
+            #
+            # To correctly gather values from flattened hyp_probs
+            # shape(hyp_probs_flat) = (batch*beam*vocabulary) we compute
+            # the offset, so the following holds:
```

Comment:
podle všecho offsetů je víc, takže kompjůtujem `offsets` a jsou to offsety těch dvojitejch pajp? jestli jo tak bych to napsal.  Starting offsets of each batch třeba.

Reply:
ping